ctr-nonce-reuse/src/ctr_game.rs

```diff
@@ -0,0 +1,196 @@
+            Self::byte_to_hex(&right_random_bytes)
+        )
+        .expect("String formatting can not fail");
+
+        flag_str
+    }
+
+    fn create_ciphertexts(&mut self, count: usize) {
+        self.ciphertexts_vec.clear();
+        let mut chosen_plaintext: Vec<_> = self
+            .plaintexts_vec
+            .choose_multiple(&mut self.rng, count)
+            .collect();
+        self.known_plaintext = (*chosen_plaintext.choose(&mut self.rng).expect("a")).clone();
+        chosen_plaintext.push(&self.flag_plaintext);
+        chosen_plaintext.shuffle(&mut self.rng);
+        for plaintext in chosen_plaintext {
+            let ciphertext = self.encrypt_bytes(plaintext);
+            self.ciphertexts_vec.push(ciphertext);
+        }
+    }
+
+    pub fn get_hex_ciphertexts(&self) -> Vec<String> {
+        let mut result = Vec::new();
+
+        for ciphertext in &self.ciphertexts_vec {
+            result.push(Self::byte_to_hex(ciphertext));
+        }
+
+        result
+    }
+
+    pub fn get_known_plaintext(&self) -> String {
+        String::from_utf8_lossy(&self.known_plaintext).to_string()
+    }
+
+    pub fn submit_flag(&self, flag_request: String) -> bool {
+        flag_request.into_bytes().eq(&self.flag_plaintext)
+    }
+
+    fn encrypt_bytes(&self, plaintext_vec: &[u8]) -> Vec<u8> {
+        let mut result = plaintext_vec.to_owned();
+
+        let mut cipher = Aes128Ctr64LE::new(&self.key.into(), &self.nonce.into());
+        cipher.apply_keystream(&mut result);
+
+        result
+    }
+
+    /*
+    fn decrypt_bytes(&self, ciphertext_vec: &[u8]) -> Vec<u8> {
+        self.encrypt_bytes(ciphertext_vec)
+    }
+    */
+
+    pub fn byte_to_hex(data: &[u8]) -> String {
+        hex::encode(data)
+    }
+
+    pub fn hex_to_text(data: String) -> String {
+        let decoded = hex::decode(data);
+        if decoded.is_err() {
+            return String::from("[unprintable]");
+        }
+        let decoded = decoded.expect("We have already handled the error");
+
+        let decoded = String::from_utf8(decoded);
+        if decoded.is_err() {
+            return String::from("[unprintable]");
+        }
+        decoded.expect("We have already handled the error")
+    }
+
+    pub fn text_to_hex(data: String) -> String {
+        hex::encode(data)
+    }
+
+    pub fn xor_hex(data: Vec<&String>) -> String {
+        let mut data_vec = Vec::new();
+        for elem in data {
+            let decoded = hex::decode(elem);
+            if decoded.is_err() {
+                return String::from("[unprintable]");
+            }
+            data_vec.push(decoded.expect("We have already handled the error"));
+        }
+        let vec_result = Self::xor_vec(data_vec);
+
+        hex::encode(vec_result)
+    }
+
+    pub fn xor_vec(data: Vec<Vec<u8>>) -> Vec<u8> {
+        let xor_result = Vec::new();
+        if data.len() < 2 {
+            return xor_result;
+        }
+        let mut xor_result = data[0].clone();
```

Comment:
To prevent allocations, we can just pop the last vector in the `data` :)
```suggestion
        let mut xor_result = data.pop().expect("length is checked");
```

Reply:
But `data` is not mutable in this function. I thought that it was a better approach not to mutate incoming variables. In case we would use this function in other places, where we need to store incoming array intact.

Reply:
It this function accepted `data: &[Vec<u8>]` then yes - the data is immutable and we have no ownership of it. 
But, this functions accepts entire `Vec<Vec<u8>>` by value. This means that the function _owns_ the vector. It means, that the function will at least deallocate it. And if you own the data, no one can prevent you from mutating it :)
In this case this will be useful, because it will save one allocation.